File: Sources/KSCrashRecording/KSCrashReportFixer.c
//
//  KSCrashReportFixer.c
//
//  Created by Karl Stenerud on 2016-11-07.
//
//  Copyright (c) 2012 Karl Stenerud. All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall remain in place
// in this source code.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//

#include <stdlib.h>
#include <string.h>

#include "KSCrashReportFields.h"
#include "KSDate.h"
#include "KSJSONCodec.h"
#include "KSLogger.h"
#include "KSSystemCapabilities.h"

#define MAX_DEPTH 100
#define MAX_NAME_LENGTH 100
#define REPORT_VERSION_COMPONENTS_COUNT 3

static const char *datePaths[][MAX_DEPTH] = {
    { "", KSCrashField_Report, KSCrashField_Timestamp },
    { "", KSCrashField_RecrashReport, KSCrashField_Report, KSCrashField_Timestamp },
};
static int datePathsCount = sizeof(datePaths) / sizeof(*datePaths);

static const char *versionPaths[][MAX_DEPTH] = {
    { "", KSCrashField_Report, KSCrashField_Version },
    { "", KSCrashField_RecrashReport, KSCrashField_Report, KSCrashField_Version },
};
static int versionPathsCount = sizeof(versionPaths) / sizeof(*versionPaths);

typedef struct {
    KSJSONEncodeContext *encodeContext;
    int reportVersionComponents[REPORT_VERSION_COMPONENTS_COUNT];
    char objectPath[MAX_DEPTH][MAX_NAME_LENGTH];
    int currentDepth;
    char *outputPtr;
    int outputBytesLeft;
} FixupContext;

static bool increaseDepth(FixupContext *context, const char *name)
{
    if (context->currentDepth >= MAX_DEPTH) {
        return false;
    }
    if (name == NULL) {
        *context->objectPath[context->currentDepth] = '\0';
    } else {
        strncpy(context->objectPath[context->currentDepth], name, sizeof(context->objectPath[context->currentDepth]));
    }
    context->currentDepth++;
    return true;
}

static bool decreaseDepth(FixupContext *context)
{
    if (context->currentDepth <= 0) {
        return false;
    }
    context->currentDepth--;
    return true;
}

static bool matchesPath(FixupContext *context, const char **path, const char *finalName)
{
    if (finalName == NULL) {
        finalName = "";
    }

    for (int i = 0; i < context->currentDepth; i++) {
        if (strncmp(context->objectPath[i], path[i], MAX_NAME_LENGTH) != 0) {
            return false;
        }
    }
    if (strncmp(finalName, path[context->currentDepth], MAX_NAME_LENGTH) != 0) {
        return false;
    }
    return true;
}

static bool matchesAPath(FixupContext *context, const char *name, const char *paths[][MAX_DEPTH], int pathsCount)
{
    for (int i = 0; i < pathsCount; i++) {
        if (matchesPath(context, paths[i], name)) {
            return true;
        }
    }
    return false;
}

static bool matchesMinVersion(FixupContext *context, int major, int minor, int patch)
{
    // Works only for report version 3.1.0 and above. See KSCrashReportVersion.h
    bool result = false;
    int *parts = context->reportVersionComponents;
    result = result || (parts[0] > major);
    result = result || (parts[0] == major && parts[1] > minor);
    result = result || (parts[0] == major && parts[1] == minor && parts[2] >= patch);
    return result;
}

static bool shouldFixDate(FixupContext *context, const char *name)
{
    return matchesAPath(context, name, datePaths, datePathsCount);
}

static bool shouldSaveVersion(FixupContext *context, const char *name)
{
    return matchesAPath(context, name, versionPaths, versionPathsCount);
}

static int onBooleanElement(const char *const name, const bool value, void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    return ksjson_addBooleanElement(context->encodeContext, name, value);
}

static int onFloatingPointElement(const char *const name, const double value, void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    return ksjson_addFloatingPointElement(context->encodeContext, name, value);
}

static int onIntegerElement(const char *const name, const int64_t value, void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    int result = KSJSON_OK;
    if (shouldFixDate(context, name)) {
        char buffer[28];

        if (matchesMinVersion(context, 3, 3, 0)) {
            ksdate_utcStringFromMicroseconds(value, buffer);
        } else {
            ksdate_utcStringFromTimestamp((time_t)value, buffer);
        }

        result = ksjson_addStringElement(context->encodeContext, name, buffer, (int)strlen(buffer));
    } else {
        result = ksjson_addIntegerElement(context->encodeContext, name, value);
    }
    return result;
}

static int onUnsignedIntegerElement(const char *const name, const uint64_t value, void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    return ksjson_addUIntegerElement(context->encodeContext, name, value);
}

static int onNullElement(const char *const name, void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    return ksjson_addNullElement(context->encodeContext, name);
}

static int onStringElement(const char *const name, const char *const value, void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    const char *stringValue = value;
    int result = ksjson_addStringElement(context->encodeContext, name, stringValue, (int)strlen(stringValue));
    if (shouldSaveVersion(context, name)) {
        memset(context->reportVersionComponents, 0, sizeof(context->reportVersionComponents));
        int versionPartsIndex = 0;
        char *mutableValue = strdup(value);
        char *versionPart = strtok(mutableValue, ".");
        while (versionPart != NULL && versionPartsIndex < REPORT_VERSION_COMPONENTS_COUNT) {
            context->reportVersionComponents[versionPartsIndex++] = atoi(versionPart);
            versionPart = strtok(NULL, ".");
        }
        free(mutableValue);
    }
    return result;
}

static int onBeginObject(const char *const name, void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    int result = ksjson_beginObject(context->encodeContext, name);
    if (!increaseDepth(context, name)) {
        return KSJSON_ERROR_DATA_TOO_LONG;
    }
    return result;
}

static int onBeginArray(const char *const name, void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    int result = ksjson_beginArray(context->encodeContext, name);
    if (!increaseDepth(context, name)) {
        return KSJSON_ERROR_DATA_TOO_LONG;
    }
    return result;
}

static int onEndContainer(void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    int result = ksjson_endContainer(context->encodeContext);
    if (!decreaseDepth(context)) {
        // Do something;
    }
    return result;
}

static int onEndData(void *const userData)
{
    FixupContext *context = (FixupContext *)userData;
    return ksjson_endEncode(context->encodeContext);
}

static int addJSONData(const char *data, int length, void *userData)
{
    FixupContext *context = (FixupContext *)userData;
    if (length > context->outputBytesLeft) {
        return KSJSON_ERROR_DATA_TOO_LONG;
    }
    memcpy(context->outputPtr, data, length);
    context->outputPtr += length;
    context->outputBytesLeft -= length;

    return KSJSON_OK;
}

char *kscrf_fixupCrashReport(const char *crashReport)
{
    if (crashReport == NULL) {
        return NULL;
    }

    KSJSONDecodeCallbacks callbacks = {
        .onBeginArray = onBeginArray,
        .onBeginObject = onBeginObject,
        .onBooleanElement = onBooleanElement,
        .onEndContainer = onEndContainer,
        .onEndData = onEndData,
        .onFloatingPointElement = onFloatingPointElement,
        .onIntegerElement = onIntegerElement,
        .onUnsignedIntegerElement = onUnsignedIntegerElement,
        .onNullElement = onNullElement,
        .onStringElement = onStringElement,
    };
    int stringBufferLength = 10000;
    char *stringBuffer = malloc((unsigned)stringBufferLength);
    int crashReportLength = (int)strlen(crashReport);
    int fixedReportLength = (int)(crashReportLength * 1.5);
    char *fixedReport = malloc((unsigned)fixedReportLength);
    KSJSONEncodeContext encodeContext;
    FixupContext fixupContext = {
        .encodeContext = &encodeContext,
        .reportVersionComponents = { 0 },
        .currentDepth = 0,
        .outputPtr = fixedReport,
        .outputBytesLeft = fixedReportLength,
    };

    ksjson_beginEncode(&encodeContext, true, addJSONData, &fixupContext);

    int errorOffset = 0;
    int result = ksjson_decode(crashReport, (int)strlen(crashReport), stringBuffer, stringBufferLength, &callbacks,
                               &fixupContext, &errorOffset);
    *fixupContext.outputPtr = '\0';
    free(stringBuffer);
    if (result != KSJSON_OK) {
        KSLOG_ERROR("Could not decode report: %s", ksjson_stringForError(result));
        free(fixedReport);
        return NULL;
    }
    return fixedReport;
}

File: Sources/KSCrashRecording/KSCrashReportFixer.m
//
//  KSCrashReportFixer.m
//
//  Created by Karl Stenerud on 2016-11-07.
//
//  Copyright (c) 2012 Karl Stenerud. All rights reserved.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall remain in place
// in this source code.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.
//

#import "KSCrashReportFixer.h"

#import <Foundation/Foundation.h>

#import "KSCrashReportFields.h"

#pragma mark - Version Parsing

static void parseVersion(NSString *versionString, int *major, int *minor, int *patch)
{
    *major = 0;
    *minor = 0;
    *patch = 0;

    if (versionString == nil) {
        return;
    }

    NSArray *parts = [versionString componentsSeparatedByString:@"."];
    if (parts.count > 0) {
        *major = [parts[0] intValue];
    }
    if (parts.count > 1) {
        *minor = [parts[1] intValue];
    }
    if (parts.count > 2) {
        *patch = [parts[2] intValue];
    }
}

static BOOL isVersionAtLeast(int major, int minor, int patch, int reqMajor, int reqMinor, int reqPatch)
{
    if (major > reqMajor) return YES;
    if (major < reqMajor) return NO;
    if (minor > reqMinor) return YES;
    if (minor < reqMinor) return NO;
    return patch >= reqPatch;
}

#pragma mark - Date Formatting

static NSString *dateStringFromMicroseconds(int64_t microseconds)
{
    // Extract seconds and microseconds parts
    time_t seconds = (time_t)(microseconds / 1000000);
    long micros = (long)(microseconds % 1000000);

    struct tm result = { 0 };
    if (gmtime_r(&seconds, &result) == NULL) {
        return nil;
    }

    // Format: yyyy-MM-ddTHH:mm:ss.xxxxxxZ (6 decimal places for microseconds)
    return [NSString stringWithFormat:@"%04d-%02d-%02dT%02d:%02d:%02d.%06ldZ", result.tm_year + 1900, result.tm_mon + 1,
                                      result.tm_mday, result.tm_hour, result.tm_min, result.tm_sec, micros];
}

static NSString *dateStringFromTimestamp(int64_t timestamp)
{
    time_t seconds = (time_t)timestamp;

    struct tm result = { 0 };
    if (gmtime_r(&seconds, &result) == NULL) {
        return nil;
    }

    // Format: yyyy-MM-ddTHH:mm:ssZ
    return [NSString stringWithFormat:@"%04d-%02d-%02dT%02d:%02d:%02dZ", result.tm_year + 1900, result.tm_mon + 1,
                                      result.tm_mday, result.tm_hour, result.tm_min, result.tm_sec];
}

#pragma mark - Fixup Logic

static void fixupTimestamp(NSMutableDictionary *reportDict, BOOL useMicroseconds)
{
    if (reportDict == nil) {
        return;
    }

    id timestamp = reportDict[KSCrashField_Timestamp];
    if ([timestamp isKindOfClass:[NSNumber class]]) {
        int64_t value = [timestamp longLongValue];
        NSString *dateString;
        if (useMicroseconds) {
            dateString = dateStringFromMicroseconds(value);
        } else {
            dateString = dateStringFromTimestamp(value);
        }
        if (dateString != nil) {
            reportDict[KSCrashField_Timestamp] = dateString;
        }
    }
}

static void fixupReport(NSMutableDictionary *report)
{
    // Get version to determine timestamp format
    id reportInfo = report[KSCrashField_Report];
    if (![reportInfo isKindOfClass:[NSDictionary class]]) {
        return;
    }
    id versionVal = reportInfo[KSCrashField_Version];

    // Determine timestamp format: 3.3.0+ writes microseconds, older writes seconds.
    // Missing or non-string version is treated as legacy (seconds) so that
    // numeric timestamps in user/custom reports are still normalized correctly.
    BOOL useMicroseconds = NO;
    if ([versionVal isKindOfClass:[NSString class]]) {
        int major, minor, patch;
        parseVersion(versionVal, &major, &minor, &patch);
        useMicroseconds = isVersionAtLeast(major, minor, patch, 3, 3, 0);
    }

    // Fix timestamp in report (mutableCopy because the input is immutable)
    NSMutableDictionary *mutableReportInfo = [reportInfo mutableCopy];
    fixupTimestamp(mutableReportInfo, useMicroseconds);
    report[KSCrashField_Report] = mutableReportInfo;

    // Fix timestamp in recrash report if present
    id recrash = report[KSCrashField_RecrashReport];
    if ([recrash isKindOfClass:[NSDictionary class]]) {
        id recrashReportInfo = recrash[KSCrashField_Report];
        if ([recrashReportInfo isKindOfClass:[NSDictionary class]]) {
            NSMutableDictionary *mutableRecrash = [recrash mutableCopy];
            NSMutableDictionary *mutableRecrashInfo = [recrashReportInfo mutableCopy];
            fixupTimestamp(mutableRecrashInfo, useMicroseconds);
            mutableRecrash[KSCrashField_Report] = mutableRecrashInfo;
            report[KSCrashField_RecrashReport] = mutableRecrash;
        }
    }
}

#pragma mark - Public API

NSDictionary *kscrf_fixupReportDict(NSDictionary *report)
{
    NSMutableDictionary *mutable = [report mutableCopy];
    fixupReport(mutable);
    return mutable;
}
